File: src/lexer.rs
use serde::{Deserialize, Serialize};

use crate::token::Token;


#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct Lexer<'a> {
    input: &'a str,
    pos: usize,
    finished: bool,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Lexer<'a> {
        Lexer { input, pos: 0, finished: false }
    }

    fn next_char(&self) -> Option<char> {
        self.input[self.pos..].chars().next()
    }

    fn starts_with(&self, s: &str) -> bool {
        self.input[self.pos..].starts_with(s)
    }

    fn read_while<F>(&mut self, test: F) -> String
    where
        F: Fn(char) -> bool,
    {
        let mut result = String::new();
        while let Some(ch) = self.next_char() {
            if !test(ch) {
                break;
            }
            result.push(ch);
            self.pos += ch.len_utf8();
        }
        result
    }
    fn read_number(&mut self) -> Token {
        // numVal: [0-9]+(\.[0-9]+)?
        // num: numVal | numVal [eE] [+-]? numVal | hexVal | octVal | binVal
        // hexVal: 0[xX] [0-9a-fA-F]+
        // octVal: 0[oO] [0-7]+
        // binVal: 0[bB] [01]+
        // so we can consume until we hit a non number
        // we can parse the number to int, float, hex, oct, bin in parser
        let s = self.read_while(|ch| {
            ch.is_ascii_digit()
                || ch == '.'
                || ch == 'e'
                || ch == 'x'
                || ch == 'o'
                || ch == 'b'
                || ch == '+'
                || ch == '-'
        });
        Token::NumberLiteral(s)
    }

    fn read_string(&mut self) -> Token {
        let s = self.read_while(|ch| ch != '"');
        self.pos += 1; // Skip the closing "
        Token::StringLiteral(s)
    }
    fn read_char(&mut self) -> Token {
        let s = self.read_while(|ch| ch != '\'' && ch != '\n');
        Token::CharLiteral(s.chars().next().unwrap())
    }

    fn read_identifier(&mut self) -> Token {
        let s = self.read_while(|ch| ch.is_ascii_alphanumeric() || ch == '_');
        Token::Identifier(s)
    }

    fn read_whitespace(&mut self) -> Token {
        // ifis newline
        let newline = self.read_while(|ch| ['\r', '\n', '\u{A0}'].contains(&ch));
        if newline.len() > 0 {
            return Token::NewLine(newline.len());
        }
        let white_space = self.read_while(|ch| ch.is_whitespace());
        if white_space.len() > 0 {
            return Token::WhiteSpace(white_space.len());
        }
        Token::WhiteSpace(0)
    }

    fn read_comment(&mut self) -> Token {
        self.pos += 2; // Skip "//"
        let comment = self.read_while(|ch| ch != '\n');
        Token::Comment(comment)
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = Token;
    

    fn next(&mut self) -> Option<Token> {
        let white_space = self.read_whitespace();
        if white_space != Token::WhiteSpace(0) {
            return Some(white_space);
        }
        match self.next_char() {
            None => {
                if self.finished {
                    return None;
                }
                self.finished = true;
                return Some(Token::EndOfFile);
            },
            Some(ch) => match ch {

                '{' => {
                    self.pos += 1;
                    Some(Token::CurlyBraceOpen)
                }
                '}' => {
                    self.pos += 1;
                    Some(Token::CurlyBraceClose)
                }
                '(' => {
                    self.pos += 1;
                    Some(Token::ParenOpen)
                }
                ')' => {
                    self.pos += 1;
                    Some(Token::ParenClose)
                }
                ':' => {
                    self.pos += 1;
                    Some(Token::Colon)
                }
                ',' => {
                    self.pos += 1;

                    Some(Token::Comma)
                }
                '"' => {
                    self.pos += 1;
                    Some(self.read_string())
                }
                '\'' => {
                    self.pos += 1;
                    Some(self.read_char())
                }
                '/' => {
                    if self.starts_with("//") {
                        Some(self.read_comment())
                    } else {
                        self.pos += 1;
                        return Some(Token::Divide);
                    }
                }
                '.' => {
                    if self.starts_with("..") {
                        self.pos += 2;
                        return Some(Token::Ellipse);
                    }
                    self.pos += 1;
                    return Some(Token::Dot);
                }
                // bitwise
                '&' => {
                    if self.starts_with("&&") {
                        self.pos += 2;
                        return Some(Token::And);
                    }
                    self.pos += 1;
                    return Some(Token::BitwiseAnd);
                }
                '|' => {
                    if self.starts_with("||") {
                        self.pos += 2;
                        return Some(Token::Or);
                    }
                    self.pos += 1;
                    return Some(Token::BitwiseOr);
                }
                '^' => {
                    self.pos += 1;
                    return Some(Token::BitwiseXor);
                }
                '~' => {
                    self.pos += 1;
                    return Some(Token::BitwiseNot);
                }
                // arithmetic
                '+' => {
                    self.pos += 1;
                    return Some(Token::Plus);
                }
                '-' => {
                    self.pos += 1;
                    return Some(Token::Minus);
                }
                '*' => {
                    self.pos += 1;
                    return Some(Token::Multiply);
                }
                '%' => {
                    self.pos += 1;
                    return Some(Token::Modulo);
                }
                // comparison
                '!' => {
                    if self.starts_with("!=") {
                        self.pos += 2;
                        return Some(Token::NotEqual);
                    }
                    self.pos += 1;
                    return Some(Token::Not);
                }
                '>' => {
                    if self.starts_with(">=") {
                        self.pos += 2;
                        return Some(Token::GreaterThanOrEqual);
                    }
                    self.pos += 1;
                    return Some(Token::GreaterThan);
                }
                '<' => {
                    if self.starts_with("<=") {
                        self.pos += 2;
                        return Some(Token::LessThanOrEqual);
                    }
                    self.pos += 1;
                    return Some(Token::LessThan);
                }
                ch if ch.is_digit(10) => Some(self.read_number()),
                ch if ch.is_ascii_alphabetic() => Some(self.read_identifier()),
                ch => panic!("Unexpected character: {}", ch),
            }
        }
    }
}

#[test]
fn test_lexer() {
    let input = r#"

myResult: MyFunction(String("Value")) {
    callBack: {
        io.print({String("Hello, world!"), Int(123)})
    },
}
"#;
    let expected = vec![
        Token::NewLine(2),
        Token::Identifier("myResult".to_string()),
        Token::Colon,
        Token::WhiteSpace(1),
        Token::Identifier("MyFunction".to_string()),
        Token::ParenOpen,
        Token::Identifier("String".to_string()),
        Token::ParenOpen,
        Token::StringLiteral("Value".to_string()),
        Token::ParenClose,
        Token::ParenClose,
        Token::WhiteSpace(1),
        Token::CurlyBraceOpen,
        Token::NewLine(1),
        Token::WhiteSpace(4),
        Token::Identifier("callBack".to_string()),
        Token::Colon,
        Token::WhiteSpace(1),
        Token::CurlyBraceOpen,
        Token::NewLine(1),
        Token::WhiteSpace(8),
        Token::Identifier("io".to_string()),
        Token::Dot,
        Token::Identifier("print".to_string()),
        Token::ParenOpen,
        Token::CurlyBraceOpen,
        Token::Identifier("String".to_string()),
        Token::ParenOpen,
        Token::StringLiteral("Hello, world!".to_string()),
        Token::ParenClose,
        Token::Comma,
        Token::WhiteSpace(1),
        Token::Identifier("Int".to_string()),
        Token::ParenOpen,
        Token::NumberLiteral("123".to_string()),
        Token::ParenClose,
        Token::CurlyBraceClose,
        Token::ParenClose,
        Token::NewLine(1),
        Token::WhiteSpace(4),
        Token::CurlyBraceClose,
        Token::Comma,
        Token::NewLine(1),
        Token::CurlyBraceClose,
        Token::NewLine(1),
        Token::EndOfFile,
    ];
    let lexer = Lexer::new(input);
    let tokens = lexer.collect::<Vec<Token>>();
    assert_eq!(tokens, expected);
}

File: src/lexer.nim
import strutils, sequtils, json

import token

type
  Lexer* = object
    input*: string
    pos*: int
    finished*: bool

proc newLexer*(input: string): Lexer =
  Lexer(input: input, pos: 0, finished: false)

proc nextChar(lex: Lexer): Option[char] =
  if lex.pos < len(lex.input):
    some(lex.input[lex.pos])
  else:
    none(char)

proc startsWith(lex: Lexer, s: string): bool =
  lex.input.startsWith(s, lex.pos)

proc readWhile(lex: var Lexer, test: proc(ch: char): bool): string =
  var result = ""
  while true:
    let ch = lex.nextChar()
    if ch.isNone or not test(ch.get()):
      break
    result.add(ch.get())
    lex.pos += 1
  result

proc readNumber(lex: var Lexer): Token =
  let s = lex.readWhile(proc(ch: char): bool =
    ch.isDigit() or ch in {'.'} or ch in ['e', 'x', 'o', 'b', '+', '-'])
  Token(kind: NumberLiteral, value: s)

proc readString(lex: var Lexer): Token =
  let s = lex.readWhile(proc(ch: char): bool = ch notin {'"', '\n'})
  lex.pos += 1
  Token(kind: StringLiteral, value: s)

proc readChar(lex: var Lexer): Token =
  let s = lex.readWhile(proc(ch: char): bool = ch notin {'\'', '\n'})
  Token(kind: CharLiteral, value: s[0])

proc readIdentifier(lex: var Lexer): Token =
  if lex.startsWith("true"):
    lex.pos += 4
    return Token(kind: BoolLiteral, value: true)
  if lex.startsWith("false"):
    lex.pos += 5
    return Token(kind: BoolLiteral, value: false)

  let s = lex.readWhile(proc(ch: char): bool = ch.isAlphaNumeric() or ch == '_')
  Token(kind: Identifier, value: s)

proc readWhitespace(lex: var Lexer): Token =
  let newline = lex.readWhile(proc(ch: char): bool = ch in {'\r', '\n', '\uA0'})
  if newline.len > 0:
    return Token(kind: NewLine, value: newline.len)

  let whiteSpace = lex.readWhile(proc(ch: char): bool = ch.isSpace())
  if whiteSpace.len > 0:
    return Token(kind: WhiteSpace, value: whiteSpace.len)

  Token(kind: WhiteSpace, value: 0)

proc readComment(lex: var Lexer): Token =
  lex.pos += 2
  let comment = lex.readWhile(proc(ch: char): bool = ch != '\n')
  Token(kind: Comment, value: comment)

iterator next*(lex: var Lexer): Token =
  let white_space = lex.readWhitespace()
  if white_space.kind != WhiteSpace or white_space.value != 0:
    yield white_space

  while true:
    case lex.nextChar()
    of None:
      if not lex.finished:
        lex.finished = true
        yield Token(kind: EndOfFile)
      else:
        break
    of some(ch):
      case ch
      of '{':
        lex.pos += 1
        yield Token(kind: CurlyBraceOpen)
      of '}':
        lex.pos += 1
        yield Token(kind: CurlyBraceClose)
      of '(':
        lex.pos += 1
        yield
        yield Token(kind: ParenOpen)
      of ')':
        lex.pos += 1
        yield Token(kind: ParenClose)
      of ':':
        lex.pos += 1
        yield Token(kind: Colon)
      of ',':
        lex.pos += 1
        yield Token(kind: Comma)
      of '"':
        lex.pos += 1
        yield lex.readString()
      of '\'':
        lex.pos += 1
        yield lex.readChar()
      of '/':
        if lex.startsWith("//"):
          yield lex.readComment()
        else:
          lex.pos += 1
          yield Token(kind: Divide)
      of '.':
        if lex.startsWith("..."):
          lex.pos += 3
          yield Token(kind: Ellipse)
        else:
          lex.pos += 1
          yield Token(kind: Dot)
      of '&':
        if lex.startsWith("&&"):
          lex.pos += 2
          yield Token(kind: And)
        else:
          lex.pos += 1
          yield Token(kind: BitwiseAnd)
      of '|':
        if lex.startsWith("||"):
          lex.pos += 2
          yield Token(kind: Or)
        else:
          lex.pos += 1
          yield Token(kind: BitwiseOr)
      of '^':
        lex.pos += 1
        yield Token(kind: BitwiseXor)
      of '~':
        lex.pos += 1
        yield Token(kind: BitwiseNot)
      of '+':
        lex.pos += 1
        yield Token(kind: Plus)
      of '-':
        lex.pos += 1
        yield Token(kind: Minus)
      of '*':
        lex.pos += 1
        yield Token(kind: Multiply)
      of '%':
        lex.pos += 1
        yield Token(kind: Modulo)
      of '!':
        if lex.startsWith("!="):
          lex.pos += 2
          yield Token(kind: NotEqual)
        else:
          lex.pos += 1
          yield Token(kind: Not)
      of '>':
        if lex.startsWith(">="):
          lex.pos += 2
          yield Token(kind: GreaterThanOrEqual)
        else:
          lex.pos += 1
          yield Token(kind: GreaterThan)
      of '<':
        if lex.startsWith("<="):
          lex.pos += 2
          yield Token(kind: LessThanOrEqual)
        else:
          lex.pos += 1
          yield Token(kind: LessThan)
      of _:
        if ch.isDigit():
          yield lex.readNumber()
        elif ch.isAlpha():
          yield lex.readIdentifier()
        else:
          raise newException(ValueError, "Unexpected character: " & $ch)
